### DISCUSSION TITLE
Replication and General Setup

### DISCUSSION THREAD

**Discussion by ProjectInitiative** [seaweedfs/seaweedfs#2312] [Q&A] [answered]
I have been learning, reading about and tinkering with seaweedfs for the past few weeks, and I figured it might be worth asking some general questions regarding basic setup.

## The plan and hardware

I have a very small cluster (3 physical machines) meant to work as a highly available cluster serving application and storage. Each of these physical servers have an 1 x SSD, and 2 x HDD. I am hoping to find a solution with seaweedfs to achieve high availability. I have been experimenting with a replication of 002 and 020 to try and achieve a 0+0+2(+1 default) = 3 copy replica. That way if I need to do maintenance on a physical machine in the cluster, I can take them down one at a time to perform the maintenance. 

## The issues

I could not be understanding the replication model, but I believe even with a 002 or 020 model, if one physical node goes offline along with all of the volume servers on that node, that I lose the ability to write to the cluster, which is not ideal for what I am trying to achieve. I have also noticed that even though I have setup 3 physical machines, each with a volume server on each machine, the filer FUSE mount still only writes data to only 1 of the 3 machines. `volume.fix.replication` and `volume.balance` don't seem to be moving volumes evenly across the cluster.

## Final outcome

I would like to know if there is an ideal setup to achieve the following, or if seaweedfs is not quite designed for this:
  * simply test out a three node cluster
  * each node has an identical SSD
  * achieve high availability
  * evenly write data across all three nodes
  * have write access to the cluster when 2/3 nodes are online

I really like what I am seeing in this project, and I would really like to get things working and contribute, so any guidance or clarification is very welcome!

**Comment by chrislusf** [comment]
SeaweedFS chooses to use strong consistency on writes. See https://github.com/chrislusf/seaweedfs/wiki/Replication#write-and-read
This means all writes needs to be successful. For your case, if there are not enough copies, the volume will be readonly.

So 3 servers are not enough for HA with 3 replicas. You would need 4 or more servers, to allow one server down.

**Reply by ProjectInitiative** [reply]
I am wondering if I can drop the replication down to 2 replicas to maintain HA with 3 servers.

**Reply by chrislusf** [reply]
Yes.

**Reply by ProjectInitiative** [reply]
Okay, thanks!

**Reply by ProjectInitiative** [reply]
> SeaweedFS chooses to use strong consistency on writes. See https://github.com/chrislusf/seaweedfs/wiki/Replication#write-and-read
> This means all writes needs to be successful. For your case, if there are not enough copies, the volume will be readonly.
> 
> So 3 servers are not enough for HA with 3 replicas. You would need 4 or more servers, to allow one server down.
> 

Re-reading this: I wanted to confirm that if I have 3 physical nodes, and a replica of 001, that the volumes placement algorithm spreads out in a way that one physical node goes down, the data is still retained, and is read/writable while the down node is brought back up. I know  gluserfs and ceph work similar to this, but instead those would usually be with a replica of 3 or higher. 

My logic is if seaweedfs uses strong consistency for writes, if you have a 3 node system with 2 total replicas, presumably the volume replicas are spread out evenly over all 3 servers. If one goes down, I'm assuming you lose write access to some of the volume replicas the offline server was hosting. 

Does this mean that those volumes are readonly unless `volume.fix.replication` completes and creates new copies on the 2 remaining servers?

Thanks again

**Reply by chrislusf** [reply]
All correct. For just 3 servers, this is fine.

Usually there should be a lot of servers.

**Reply by ProjectInitiative** [reply]
This means that true high availability can not be achieved then? I am coming to this conclusion because of the following constraints:

  * If there are more volume servers or physical servers, that doesn't change the replica count, it would still remain the same, just more spread out. Meaning the same problem from before, where if one (or many) servers die, those volumes have to be replicated again before being fully operational.

* There would be a down time of full capabilities of the volumes until either the server(s) is brought back up, or volumes replicated are complete.

However, reading the replication wiki:
W + R > N 
If W = N, R = 1 and I define N to be 001, the system reads that as 0+0+1 (+1 by default) = 2 copies. N = 2 in this case. 

2 + 1 > 2 
If I lose one replica in any way, that volumes quorum drops to 1 + 1 > 2 which is false therefore not writable until corrected. 

I'd argue that this just keeps data safe, but isn't highly available, as highly available would mean  I could increase my replica count to anything as long as don't lose more than half of the total number where the quorum would be: 
W + R >= N
Or even better: W + R >= N/2

Please let me know thoughts, as this is how I currently understand the functionality :)

**Reply by chrislusf** [reply]
* For new writes, pick a different set of replicas.
* Leave existing data as is. If you want to make existing data writable and make them HA, add back the lost replica or use `volume.fix.replication` to make another copy. That depends on why the replica was missing. If it is a transient failure, just wait it out. If it is a permanent problem, need to fix the replication.

**Reply by ProjectInitiative** [reply]
Okay, that makes more sense, new writes just get moved to new volumes. 

Main reason fro bringing this up was because of there was a VM disk running and the virtual disk file is chunked or spread out and the situation described above is transient, then the VM could get stuck in the meantime. Thank you for the clarification, I believe I have a better understanding of the underlying architecture

**Comment by chrislusf** [comment]
> the filer FUSE mount still only writes data to only 1 of the 3 machines

Need to see how and where you configured the replication.

**Reply by ProjectInitiative** [reply]
I added the `-defaultReplication=002` to the master servers. I was reading that there is a hierarchy for replication rules: FUSE, filer, and then master. If that is the case, I am wondering if I would need to specify the replication for each mount to the filer. i.e. every client machine would need to include a replication scheme. 

Is there a way to have a global rule override that ignores all client settings?

**Reply by chrislusf** [reply]
> Is there a way to have a global rule override that ignores all client settings?

No. The client settings take priority.

**Reply by ProjectInitiative** [reply]
Okay, that probably solves the replication issue, I'll retry with client settings!

**Reply by ProjectInitiative** [reply]
I dropped the master down to a replica of 001 and left the filer and client blank so it should be using the master's replication settings if I read the help menus right. I still am only seeing data written to the first volume server and not spread out. 

Further more when running `volume.fix.replication` multiple volumes are duplicated on the same volume server and not spread out. Not really sure what I should check next

**Reply by chrislusf** [reply]
show your setup.

**Reply by ProjectInitiative** [reply]
[setup.zip](https://github.com/chrislusf/seaweedfs/files/7118839/setup.zip)

Here is everything that I have setup. filer.toml shared, docker swarm yaml file. 

Essentially I have three physical nodes (node01, node02, node03) that are apart of a docker swarm. 
Running the commands in the `commands.sh` file sets everything up given that the `filer.toml` are all placed on the nodes. The end result is as I described above, and the FUSE client tends to hang on any uploads. All volumes seem to get created and replicated only on the first server (see screenshot).
![Screenshot from 2021-09-06 23-20-15](https://user-images.githubusercontent.com/6314611/132283795-70bcafd5-3857-4966-acb4-3f9f6a5795a4.png)

**Reply by chrislusf** [reply]
what are the outputs of `volume.list` and `volume.balance -force`?

**Reply by ProjectInitiative** [reply]
```
> volume.list
Topology volumeSizeLimit:1024 MB hdd(volume:13/900 active:13 free:887 remote:0)
  DataCenter dc1 hdd(volume:13/900 active:13 free:887 remote:0)
    Rack rack1 hdd(volume:13/900 active:13 free:887 remote:0)
      DataNode 10.0.0.26:9991 hdd(volume:13/300 active:13 free:287 remote:0)
        Disk hdd(volume:13/300 active:13 free:287 remote:0)
          volume id:2 size:8 replica_placement:1 version:3 modified_at_second:1630968011
          volume id:5 size:8 replica_placement:1 version:3 modified_at_second:1630968075
          volume id:6 size:8 replica_placement:1 version:3 modified_at_second:1630968097
          volume id:7 size:8 replica_placement:1 version:3 modified_at_second:1630968118
          volume id:11 size:8 replica_placement:1 version:3 modified_at_second:1630968205
          volume id:13 size:8 replica_placement:1 version:3 modified_at_second:1630968248
          volume id:22 size:8 replica_placement:1 version:3 modified_at_second:1630968441
          volume id:23 size:8 replica_placement:1 version:3 modified_at_second:1630968463
          volume id:26 size:8 replica_placement:1 version:3 modified_at_second:1630968527
          volume id:27 size:8 replica_placement:1 version:3 modified_at_second:1630968549
          volume id:28 size:8 replica_placement:1 version:3 modified_at_second:1630968570
          volume id:30 size:8 replica_placement:1 version:3 modified_at_second:1630968613
          volume id:31 size:8 replica_placement:1 version:3 modified_at_second:1630968635
        Disk hdd total size:104 file_count:0
      DataNode 10.0.0.26:9991 total size:104 file_count:0
      DataNode 10.0.0.30:9993 hdd(volume:0/300 active:0 free:300 remote:0)
        Disk hdd(volume:0/300 active:0 free:300 remote:0)
        Disk hdd total size:0 file_count:0
      DataNode 10.0.0.30:9993 total size:0 file_count:0
      DataNode 10.0.2.208:9992 hdd(volume:0/300 active:0 free:300 remote:0)
        Disk hdd(volume:0/300 active:0 free:300 remote:0)
        Disk hdd total size:0 file_count:0
      DataNode 10.0.2.208:9992 total size:0 file_count:0
    Rack rack1 total size:104 file_count:0
  DataCenter dc1 total size:104 file_count:0
total size:104 file_count:0

> lock
> volume.balance -force
hdd 0.01 0.04:0.00      hdd 0.01 0.04:0.00     
>
```

**Comment by chrislusf** [comment]
what is the output for `volume.fix.replication`?

**Reply by ProjectInitiative** [reply]
```
> lock
> volume.fix.replication
replicating volume 7 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 6 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 31 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 28 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 26 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 2 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 11 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 13 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 5 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 23 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 30 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 27 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
replicating volume 22 001 from 10.0.0.50:9991 to dataNode 10.0.2.81:9993 ...
error: copying from 10.0.0.50:9991 => 10.0.2.81:9993 : rpc error: code = Unknown desc = read volume file status failed, rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.0.0.50:19991: i/o timeout"
```
Seems to fail at the end, No change in the volume placements.

**Reply by chrislusf** [reply]
seems there is a network problem. Volume server on `10.0.2.81:9993` is trying to read from `10.0.0.50:19991`. I could not tell where is the setup issue.

After fixing the replication, the volume balance should work.

**Reply by ProjectInitiative** [reply]
Should the volume server on 10.0.2.81:9993 be reading from 10.0.0.50:9991 instead of 10.0.0.50:19991? Otherwise I'm not sure what other networking issues there would be as all ports specified are open, and I can connect to the specified ports via telnet.

Thanks!

**Reply by ProjectInitiative** [reply]
> seems there is a network problem. Volume server on `10.0.2.81:9993` is trying to read from `10.0.0.50:19991`. I could not tell where is the setup issue.
> 
> After fixing the replication, the volume balance should work.

Regarding this issue: I still was unable to figure out why the gRPC connection was failing withing Docker swarms overlay network. Testing externally things work as expected. After doing some research, some possible reasons for the timeout could be due to there being too many connections to the gRPC client because of how Docker overlay network load balances, or the timeout isn't long enough. Apart from that I don't really have any clue why the docker swarm networking breaks.

**Comment by ProjectInitiative** [comment, marked as answer]
Alright so the solution to all the problems was to not use Docker in any way shape or form. Set everything up on bare metal and things are working and distributing as expected. I am not sure why Docker's networking is not working. :/

**Reply by chrislusf** [reply]
Pretty sure it is your docker setup problem on network.

**Reply by ProjectInitiative** [reply]
The way I am looking at it is if the cluster runs fine on the network on bare metal, and docker is added completely original with no networking, or firewalls added then it piggy backs on the bare metal network (which already proved to work). I understand the sentiment of thinking there is a docker problem with the network. I was just trying to express that this might be an issue with docker and not the setup as I have found a couple of other gRPC issues with vanilla docker installations. :) 
Also want to mention that Docker Swarm overlay networks do work differently than the standard ones too.
